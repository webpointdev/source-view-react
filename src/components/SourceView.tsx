import {useMemo, FC, ReactNode} from 'react';
import {LineOfSyntax, SyntaxElement, TreeNode} from 'source-tokenizer';

export type RenderSyntaxTree = (root: TreeNode, defaultRender: RenderSyntaxElement, i: number) => ReactNode;

export type RenderSyntaxElement = (element: SyntaxElement, i: number) => ReactNode;

interface TextViewProps {
    source: string;
}

interface SyntaxViewProps {
    syntax: LineOfSyntax[];
    renderSyntaxTree?: RenderSyntaxTree;
}

export type SourceViewProps = TextViewProps & SyntaxViewProps;

const renderGenericElement = (element: SyntaxElement, i: number): ReactNode => {
    if (typeof element === 'string') {
        return element;
    }

    return (
        <span key={i} className={element.properties?.className?.join(' ')}>
            {element.children.map(renderGenericElement)}
        </span>
    );
};

const renderGutter = (line: number): ReactNode => <td className="source-gutter" data-line-number={line} />;

const renderSourceContent = (source: string): ReactNode => <td className="source-code">{source}</td>;

const renderSyntaxContent = (syntax: LineOfSyntax, {renderSyntaxTree}: SourceViewProps): ReactNode => {
    const render = (element: SyntaxElement, i: number) => {
        if (typeof element === 'string') {
            return element;
        }

        if (renderSyntaxTree) {
            return renderSyntaxTree(element, renderGenericElement, i);
        }

        return renderGenericElement(element, i);
    };

    return <td className="source-code">{syntax.map(render)}</td>;
};

const renderSourceLine = (content: string, i: number): ReactNode => (
    <tr key={i} className="source-line">
        {renderGutter(i + 1)}
        {renderSourceContent(content)}
    </tr>
);

const renderSyntaxLine = (content: LineOfSyntax, i: number, props: SourceViewProps): ReactNode => (
    <tr key={i} className="source-line">
        {renderGutter(i + 1)}
        {renderSyntaxContent(content, props)}
    </tr>
);

export const SourceView: FC<SourceViewProps> = props => {
    const {source, syntax} = props;
    const lines = useMemo(() => source?.split('\n') ?? [], [source]);

    return (
        <table className="source">
            <tbody>
                {
                    syntax
                        ? syntax.map((item, i) => renderSyntaxLine(item, i, props))
                        : lines.map(renderSourceLine)
                }
            </tbody>
        </table>
    );
};
