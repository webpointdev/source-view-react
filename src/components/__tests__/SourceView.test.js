/* eslint-disable react/jsx-no-bind */
import {render} from '@testing-library/react';
import {highlight} from 'refractor';
import {tokenize} from 'source-tokenizer';
import {SourceView} from '../SourceView';

const source = `const a = 3;
console.log(3);`;

describe('SourceView', () => {
    test('text source', () => {
        const {asFragment} = render(<SourceView source={source} />);
        expect(asFragment()).toMatchSnapshot();
    });

    test('syntax source', () => {
        const options = {
            highlight(source) {
                return highlight(source, 'javascript');
            },
        };
        const syntax = tokenize(source, options);
        const {asFragment} = render(<SourceView source={source} syntax={syntax} />);
        expect(asFragment()).toMatchSnapshot();
    });

    test('custom syntax renderer', () => {
        const options = {
            highlight(source) {
                return highlight(source, 'javascript');
            },
        };
        const syntax = tokenize(source, options);
        const stripSyntax = node => {
            if (typeof node === 'string') {
                return node;
            }

            return node.children.map(stripSyntax);
        };
        const {asFragment} = render(<SourceView syntax={syntax} renderSyntaxTree={stripSyntax} />);
        expect(asFragment()).toMatchSnapshot();
    });
});
