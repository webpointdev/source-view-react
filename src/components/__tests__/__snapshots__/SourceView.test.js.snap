// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SourceView custom syntax renderer 1`] = `
<DocumentFragment>
  <table
    class="source"
  >
    <tbody>
      <tr
        class="source-line"
      >
        <td
          class="source-gutter"
          data-line-number="1"
        />
        <td
          class="source-code"
        >
          const a = 3;
        </td>
      </tr>
      <tr
        class="source-line"
      >
        <td
          class="source-gutter"
          data-line-number="2"
        />
        <td
          class="source-code"
        >
          console.log(3);
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`SourceView syntax source 1`] = `
<DocumentFragment>
  <table
    class="source"
  >
    <tbody>
      <tr
        class="source-line"
      >
        <td
          class="source-gutter"
          data-line-number="1"
        />
        <td
          class="source-code"
        >
          <span
            class="token keyword"
          >
            const
          </span>
           a 
          <span
            class="token operator"
          >
            =
          </span>
           
          <span
            class="token number"
          >
            3
          </span>
          <span
            class="token punctuation"
          >
            ;
          </span>
        </td>
      </tr>
      <tr
        class="source-line"
      >
        <td
          class="source-gutter"
          data-line-number="2"
        />
        <td
          class="source-code"
        >
          <span
            class="token console class-name"
          >
            console
          </span>
          <span
            class="token punctuation"
          >
            .
          </span>
          <span
            class="token method function property-access"
          >
            log
          </span>
          <span
            class="token punctuation"
          >
            (
          </span>
          <span
            class="token number"
          >
            3
          </span>
          <span
            class="token punctuation"
          >
            )
          </span>
          <span
            class="token punctuation"
          >
            ;
          </span>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`SourceView text source 1`] = `
<DocumentFragment>
  <table
    class="source"
  >
    <tbody>
      <tr
        class="source-line"
      >
        <td
          class="source-gutter"
          data-line-number="1"
        />
        <td
          class="source-code"
        >
          const a = 3;
        </td>
      </tr>
      <tr
        class="source-line"
      >
        <td
          class="source-gutter"
          data-line-number="2"
        />
        <td
          class="source-code"
        >
          console.log(3);
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
